src/dataflow-types/src/client.rs

```diff
@@ -283,57 +311,287 @@ pub trait Client: Send {
+            command.frontier_tracking(&mut start, &mut cease);
+            // Apply the determined effects of the command to `self.uppers`.
+            for id in start.into_iter() {
+                let mut frontier = timely::progress::frontier::MutableAntichain::new();
+                frontier.update_iter(Some((
+                    <Timestamp as timely::progress::Timestamp>::minimum(),
+                    self.parts as i64,
+                )));
+                let previous = self.uppers.insert(id, frontier);
+                assert!(previous.is_none());
+            }
+            for id in cease.into_iter() {
+                let previous = self.uppers.remove(&id);
+                assert!(previous.is_some());
+            }
+        }
+
+        /// Absorbs a response, and produces response that should be emitted.
+        pub fn absorb_response(&mut self, shard_id: usize, message: Response) -> Option<Response> {
+            match message {
+                Response::FrontierUppers(mut list) => {
+                    // Fold updates into the maintained antichain, and report
+                    // any net changes to the minimal antichain itself.
+                    let mut reactions = ChangeBatch::new();
+                    for (id, changes) in list.iter_mut() {
+                        // We may receive changes to identifiers that are no longer tracked;
+                        // do not worry about them in that case (a benign race condition).
+                        if let Some(frontier) = self.uppers.get_mut(id) {
+                            for (time, diff) in frontier.update_iter(changes.drain()) {
+                                reactions.update(time, diff);
+                            }
+                            std::mem::swap(changes, &mut reactions);
+                        } else {
+                            changes.clear();
+                        }
+                    }
+                    // TODO: The following line would be great, but is not permitted by `list.retain()`.
+                    // list.retain_mut(|(_, changes)| !changes.is_empty());
+                    if !list.is_empty() {
+                        Some(Response::FrontierUppers(list))
+                    } else {
+                        None
+                    }
+                }
+                Response::PeekResponse(connection, response) => {
+                    // Incorporate new peek responses; awaiting all responses.
+                    let entry = self
+                        .peek_responses
+                        .entry(connection)
+                        .or_insert_with(|| Default::default());
+                    let novel = entry.insert(shard_id, response);
+                    assert!(novel.is_none(), "Duplicate peek response");
+                    // We may be ready to respond.
+                    if entry.len() == self.parts {
+                        let mut response = PeekResponse::Rows(Vec::new());
```

Comment:
Nit: Compute the size of `rows` and pre-allocate?

Reply:
I think there is an even smarter thing to do, where you pre-check all the responses to see if any are error or cancelations, and maybe even smarter to maintain that state rather than a vector of responses (i.e. drop rows as soon as you know you won't need them). I might just add a TODO, though.

Reply:
Or even better, allow the responses to be chunked as `Vec<Vec<_>>` which would allow us to just appropriate the allocations, and is also more frame-size friendly!